pkg/controller/cluster/cluster_controller_phases.go

```diff
@@ -0,0 +1,195 @@
+	obj, err := external.Get(r.Client, ref, cluster.Namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) && !cluster.DeletionTimestamp.IsZero() {
+			return nil, nil
+		} else if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
+				"could not find %v %q for Cluster %q in namespace %q, requeuing",
+				ref.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return nil, err
+	}
+
+	objPatch := client.MergeFrom(obj.DeepCopy())
+
+	// Delete the external object if the Cluster is being deleted.
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.Delete(ctx, obj); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to delete %v %q for Cluster %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return obj, nil
+	}
+
+	// Set external object OwnerReference to the Cluster.
+	ownerRef := metav1.OwnerReference{
+		APIVersion: cluster.APIVersion,
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !util.HasOwnerRef(obj.GetOwnerReferences(), ownerRef) {
+		obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
+		if err := r.Patch(ctx, obj, objPatch); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to set OwnerReference on %v %q for Cluster %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+	}
+
+	// Add watcher for external object, if there isn't one already.
+	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
+	if !loaded && r.controller != nil {
+		klog.Infof("Adding watcher on external object %q", obj.GroupVersionKind())
+		err := r.controller.Watch(
+			&source.Kind{Type: obj},
+			&handler.EnqueueRequestForOwner{OwnerType: &clusterv1.Cluster{}},
+		)
+		if err != nil {
+			r.externalWatchers.Delete(obj.GroupVersionKind().String())
+			return nil, errors.Wrapf(err, "failed to add watcher on external object %q", obj.GroupVersionKind())
+		}
+	}
+
+	// Set error reason and message, if any.
+	errorReason, errorMessage, err := external.ErrorsFrom(obj)
+	if err != nil {
+		return nil, err
+	}
+	if errorReason != "" {
+		clusterStatusError := capierrors.ClusterStatusError(errorReason)
+		cluster.Status.ErrorReason = &clusterStatusError
+	}
+	if errorMessage != "" {
+		cluster.Status.ErrorMessage = pointer.StringPtr(errorMessage)
+	}
+
+	return obj, nil
+}
+
+// reconcileInfrastructure reconciles the Spec.InfrastructureRef object on a Cluster.
+func (r *ReconcileCluster) reconcileInfrastructure(ctx context.Context, cluster *v1alpha2.Cluster) error {
+	if cluster.Spec.InfrastructureRef == nil {
+		return nil
+	}
+
+	// Call generic external reconciler.
+	infraConfig, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
+	if infraConfig == nil && err == nil {
```

Comment:
Would you be willing to stipulate that if err is nil, the returned object will always be non-nil? If so, then:
```go
if err != nil {
  return err
}
// proceed
```

Reply:
I can follow-up in another PR and change the Machine part as well. It definitely needs some clean up

Reply:
The change you're suggesting would make us return a predefined error which callers will need to check, does that sound good?

Reply:
Why would that necessitate a predefined error?

Reply:
Discussed on Zoom. Going to refactor to make this cleaner in a follow-up PR.